### PR TITLE
fix nested no_grad decorator and with-statement

### LINKED ISSUE #10858: unexpected behaviour when mixing no_grad decorator with no_grad context manager
## Issue description

Mixing torch.no_grad decorator with torch.no_grad context re-enables gradients as per torch.is_grad_enabled.

## Code example

```
import torch

@torch.no_grad()
def nothing(x):
    return x

testin = torch.Tensor([0])
with torch.no_grad():
    print(torch.is_grad_enabled())
    testout = nothing(testin)
    print(torch.is_grad_enabled())
```

### Expectation

It would be my expectation that within the no_grad context the gradients would always be disabled.
This expectation matches the output when using nested contexts as below.

```
import torch

def nothing_no_nograd(x):
  return x

testin = torch.Tensor([0])
with torch.no_grad():
    print(torch.is_grad_enabled())
    with torch.no_grad():
        testout = nothing_no_nograd(testin)
    print(torch.is_grad_enabled())
```

## System Info

PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.2.148

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 396.54
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] Could not collect
[conda] cuda92                    1.0                           0    pytorch
[conda] pytorch                   0.4.1           py36_cuda9.2.148_cudnn7.1.4_1  [cuda92]  pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch
[conda] warpctc-pytorch           0.1                       <pip>

### PR DESCRIPTION
**Pull request by weiyangfb**
- fixes https://github.com/pytorch/pytorch/issues/10858
- allow `no_grad` decorator to apply `with torch.no_grad()` at the correct context
- current behavior:
```
import torch

@torch.no_grad()
def nothing(x):
    return x

testin = torch.Tensor([0])
with torch.no_grad():
    print(torch.is_grad_enabled()) # False
    testout = nothing(testin)
    print(torch.is_grad_enabled()) # False
```

### CODE REVIEW

File: torch/autograd/grad_mode.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def __exit__(self, *args):
     def __call__(self, func):
         @functools.wraps(func)
         def decorate_no_grad(*args, **kwargs):
-            with self:
+            with torch.no_grad():
```
Comment (SsnL) [inline torch/autograd/grad_mode.py]: I am very confused. Why did `with self` not work?

Comment (weiyangfb) [reply torch/autograd/grad_mode.py]: Yeah, I don't understand either, but here is the difference I saw:

Using `with self`:
```
class Decor():
    def __init__(self):
        print('__init__ called')
        self.init_var = 0
        
    def __enter__(self):
        print('__enter__ called')
        return self
    
    def __exit__(self, exc_type, exc_value, exc_traceback):
        print('__exit__ called')
    
    def __call__(self, func):
        def decorate(*args, **kwargs):
            with self:
                return func(*args, **kwargs)
        return decorate

@Decor()
def foo():
    print('inside with statement body')
print("==============")
foo()
######################
__init__ called
==============
__enter__ called
inside with statement body
__exit__ called
```

Using `with Decor()`:
```
class Decor():
    def __init__(self):
        print('__init__ called')
        self.init_var = 0
        
    def __enter__(self):
        print('__enter__ called')
        return self
    
    def __exit__(self, exc_type, exc_value, exc_traceback):
        print('__exit__ called')
    
    def __call__(self, func):
        def decorate(*args, **kwargs):
            with self:
                return func(*args, **kwargs)
        return decorate

@Decor()
def foo():
    print('inside with statement body')
print("==============")
foo()
###################
__init__ called
==============
__init__ called
__enter__ called
inside with statement body
__exit__ called
```

Notice `__init__` gets invoked once more when `with Decor()` decorated function is called

**Review by apaszke — APPROVED** [review]

File: torch/autograd/grad_mode.py (hunk 1 above)
Comment (SsnL) [reply torch/autograd/grad_mode.py]: Yes this seems expected, because when you do `with Decor()`, a new instance is created. I will think more why `with self` doesn't work.

Comment (weiyangfb) [reply torch/autograd/grad_mode.py]: is it sufficient if we know that `self` only represents the class object, and thus with `with self:` context only gets recorded once when decorator applied on a function, where if we use `with DecoratorName()`, a new instance gets created whenever the decorated function is called, and so the latest context will be recorded.

Comment (SsnL) [reply torch/autograd/grad_mode.py]: Okay I understand now. Recall the original bahavior:
```
import torch

@torch.no_grad()
def nothing(x):
    return x

testin = torch.Tensor([0])
with torch.no_grad():
    print(torch.is_grad_enabled()) # False
    testout = nothing(testin)
    print(torch.is_grad_enabled()) # True
```

This happens, because we save `self.prev = torch.is_grad_enabled()` in `__init__` not in `__enter__`. So when the `no_grad` object is created in `@torch.no_grad()` , it saves `self.prev = True`. And then after `nothing`, the `grad_mode` is restored to `True` at its `__exit__`. To fix this. We should just make `__init__` a no-op, and save the `self.prev` in `__enter__`.

**Review by ssnl — CHANGES_REQUESTED** [review]
See comment for a more proper fix

File: torch/autograd/grad_mode.py (hunk 1 above)
Comment (weiyangfb) [reply torch/autograd/grad_mode.py]: Thanks! I will change it accordingly

**Review by ssnl — APPROVED** [review]
